tests/describe_task_queue_test.go

```diff
@@ -0,0 +1,230 @@
+
+	tl := "backlog-counter-task-queue"
+	tq := &taskqueuepb.TaskQueue{Name: tl, Kind: enumspb.TASK_QUEUE_KIND_NORMAL}
+	identity := "worker-multiple-tasks"
+
+	for i := 0; i < workflows; i++ {
+
+		id := uuid.New()
+		wt := "functional-workflow-multiple-tasks"
+		workflowType := &commonpb.WorkflowType{Name: wt}
+
+		request := &workflowservice.StartWorkflowExecutionRequest{
+			RequestId:           uuid.New(),
+			Namespace:           s.namespace,
+			WorkflowId:          id,
+			WorkflowType:        workflowType,
+			TaskQueue:           tq,
+			Input:               nil,
+			WorkflowRunTimeout:  durationpb.New(20 * time.Second),
+			WorkflowTaskTimeout: durationpb.New(3 * time.Second),
+			Identity:            identity,
+		}
+
+		resp0, err0 := s.engine.StartWorkflowExecution(NewContext(), request)
+		s.NoError(err0)
+
+		we := &commonpb.WorkflowExecution{
+			WorkflowId: id,
+			RunId:      resp0.RunId,
+		}
+
+		s.EqualHistoryEvents(`
+  1 WorkflowExecutionStarted
+  2 WorkflowTaskScheduled`, s.getHistory(s.namespace, we))
+	}
+
+	response := s.ValidateDescribeTaskQueue(tl)
+	versionInfo := response.GetVersionsInfo()[""]
+
+	// verifying backlog counter increased as tasks were added
+	addedTasks := int64(workflows)
+	for qT, t := range versionInfo.GetTypesInfo() {
+		queueType := enumspb.TaskQueueType(qT)
+		if queueType == enumspb.TASK_QUEUE_TYPE_WORKFLOW {
+			s.Assert().Equal(t.BacklogInfo.ApproximateBacklogCount, addedTasks)
+		} else if queueType == enumspb.TASK_QUEUE_TYPE_ACTIVITY {
+			s.Assert().Equal(t.BacklogInfo.ApproximateBacklogCount, int64(0))
+		}
+	}
+
+	// starting and completing the workflow tasks by fetching their workflowExecutions
+	for i := 0; i < workflows; i++ {
+		resp1, err1 := s.engine.PollWorkflowTaskQueue(NewContext(), &workflowservice.PollWorkflowTaskQueueRequest{
+			Namespace: s.namespace,
+			TaskQueue: tq,
+			Identity:  identity,
+		})
+		s.NoError(err1)
+		s.Equal(int32(1), resp1.GetAttempt())
+
+		// responding with completion
+		_, err2 := s.engine.RespondWorkflowTaskCompleted(NewContext(), &workflowservice.RespondWorkflowTaskCompletedRequest{
+			Namespace:             s.namespace,
+			TaskToken:             resp1.GetTaskToken(),
+			ReturnNewWorkflowTask: true,
+		})
+		s.NoError(err2)
+	}
+
+	// call describeTaskQueue to verify if the backlog decreased (should decrease since we only had one task)
+	response = s.ValidateDescribeTaskQueue(tl)
+	versionInfo = response.GetVersionsInfo()[""]
+
+	for _, t := range versionInfo.GetTypesInfo() {
+		s.Assert().Equal(t.BacklogInfo.ApproximateBacklogCount, int64(0))
+	}
+}
+
+func (s *DescribeTaskQueueSuite) ValidateDescribeTaskQueue(tl string) *workflowservice.DescribeTaskQueueResponse {
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+
+	var resp *workflowservice.DescribeTaskQueueResponse
+	var err error
+
+	s.Eventually(func() bool {
```

Comment:
does this`Eventually` do anything if it's not waiting for the backlog info to eventually contain the expected counts?

I'd say either consider calling DescribeTaskQueue just once without Eventually in here and rename the helper function to something like `callDescribeTaskQueue` (because it's not doing much validation as it is now).

OR, you can rename it to something like `validateBacklogCount` but also accept a `expectedBacklogCount map[enumspb.TaskQueueType][int64]` in the args and keep `Eventually` waiting to eventually see the expected counts.

Second option will result in a less flaky test because it accounts for the counter delay due to transfer task processing.

[In both cases you probably want to keep the method private, not starting with a capital letter]

Reply:
I think the second approach is better since it's more fault tolerant, yes. Implemented.